### PR TITLE
knex transaction usage

### PR DESCRIPTION
**Pull request by darahayes**
Shows how transactions can be used during conflict detection/resolution.

### CODE REVIEW

File: server-apollo/src/schema.js [hunk 1]
```diff
@@ -49,18 +49,22 @@ const resolvers = {
       return result
     },
     updateUser: async (obj, args, context, info) => {
-      let { id, version, ...updateArgs } = args
-      const currentRecord = await context.db('users').select().where('id', args.id).then((rows) => rows[0])
-      if (!currentRecord) return null // or not found error??
+      return new Promise((resolve, reject) => {
```
Comment (wtrocki) [inline server-apollo/src/schema.js:52]: Let's move this to SDK as separate resolve and merge it so we will have ready to go example for our own server

File: server-apollo/src/schema.js [hunk 2]
```diff
@@ -49,18 +49,22 @@ const resolvers = {
...
-      if (conflict) {
-        updateArgs = context.handleConflict(context.conflictHandlers.RETURN_TO_CLIENT, conflict, currentRecord, args)
-      }
+          if (conflict) {
+            updateArgs = context.handleConflict(context.conflictHandlers.RETURN_TO_CLIENT, conflict, currentRecord, args)
+          }
 
-      const result = await context.db('users').update({ ...updateArgs, version: currentRecord.version + 1 }).where({ 'id': id }).returning('*').then((rows) => rows[0])
-      return result
+          const result = await trx('users').update({ ...updateArgs, version: currentRecord.version + 1 }).where({ 'id': id }).returning('*').then((rows) => rows[0])
```
Comment (wtrocki) [inline server-apollo/src/schema.js:64]: trx can be still context.db in this case. Is it possible to execute the same operations on `context.db` as on `trx` ?

Comment (darahayes) [reply server-apollo/src/schema.js:64]: I don't think so. I'm guessing that whatever happens underneath is that the `trx` thing references back to the transaction. However, we would probably open the transaction and then provide the user resolver with the trx object :)

File: server-apollo/src/schema.js (hunk 1 above)
Comment (darahayes) [reply server-apollo/src/schema.js:52]: +1 good idea

File: server-apollo/src/schema.js (hunk 2 above)
Comment (wtrocki) [reply server-apollo/src/schema.js:64]: So, put some thoughts into this. However users need to be able to perform select and update within transaction which is happening here. So ideally we should just expose trx as method and document that as knex transation.

**Comment by wtrocki** [comment]
We missing solution for the problem when version is missing (not supported/supplied) but we still provide transaction support. IMHO we may have top level resolver for detection/enablement that will create and supply transaction as separate element of the context. Alternatively we can leave transactions up to users and have all code inside. Then processing is simple and it can be documented as:

- Initiate transaction
- **Perform read (can be multiple read operations)**
- Execute conflict helper logic with one of strategies
- **Perform write on db**
- Close transaction

## Wrapper idea
We could have something like Conflict Resolver that will allow users to supply:
Read operation
Write operation
Conflict strategy 

This will simplify logic and allow us to handle internal parts of the process without forcing users to rewrite their resolvers.
```javascript
  withConflict({
    read: (params) = {
      // Read operation
    },
    write: (data) => {
      // Write operation
    },
    strategy: strategies.CLIENT
  }
```

This will mean that trx will be supported inside specific use case and it will be detected when needed.
This will drastically simplify documentation and business logic and will allow us to avoid problems with the 
transactions setup for resolvers that have no conflict stuff.

**Review by wtrocki — APPROVED** [review]
Reviewed. Really good job!

**Comment by darahayes** [comment]
@wtrocki Going to merge this PR as it is and then I'm going to prototype this wrapper idea as a separate PR 😃

**Pull request merged**